corehq/apps/hqwebapp/forms.py

```diff
@@ -291,12 +291,17 @@ def as_table(self):
 
 
 class HQAuthenticationTokenForm(AuthenticationTokenForm):
+    def __init__(self, user, initial_device, request, **kwargs):
+        super().__init__(user, initial_device, **kwargs)
+        self.request = request
 
     def clean(self):
         try:
             cleaned_data = super(HQAuthenticationTokenForm, self).clean()
         except ValidationError:
-            user_login_failed.send(sender=__name__, credentials={'username': self.user.username})
+            user_login_failed.send(sender=__name__, credentials={'username': self.user.username},
+                request=self.request,
+                token_failure=True)
```

Comment:
is it safe to add this argument? It's not one of the ones listed in the Django docs.

Reply:
We want to be able to distinguish between a generic login failure and one caused by a failed token. Given that both result in the user failing to log in, I think re-using this signal is appropriate -- the alternative is creating an entirely new signal specifically for token failures.

As for safety, I think this is safe. [The docs](https://docs.djangoproject.com/en/3.1/topics/signals/#receiver-functions) mention: `All signals send keyword arguments, and may change those keyword arguments at any time...at any point arguments could get added to the signal and your receiver must be able to handle those new arguments`. So, `token_failure` should get processed as a kwarg if the receiver doesn't recognize it. Do you fear adding a custom parameter like this would be confusing to someone unfamiliar with the code?